File: appli-android/app/src/main/java/com/example/sportapp/core/data/model/HealthStepCount.kt
package com.example.sportapp.core.data.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey
import com.example.sportapp.core.utils.CustomDateUtils.getNowISO8601

/**
 * Santé / Health Connect V1 (2026-06-17) — portage Android du modèle serveur
 * `health_step_counts` (cf. serveur/app/models/health_step_count.py). Type A
 * user-scoped.
 *
 * Compteur de pas en buckets intraday : chaque row = une tranche de la journée
 * (bucket horaire). Le total quotidien se dérive par SUM(steps) sur la date.
 * L'`uuid` est stable par user+date+bucketStart (cf. HealthImporter) : le bucket
 * courant est ré-upsert au fil de la journée sans créer de doublon.
 */
@Entity(
    tableName = "health_step_counts",
    indices = [
        Index(value = ["uuid"], unique = true),
        Index(value = ["user_id"]),
        Index(value = ["date"]),
    ]
)
data class HealthStepCount(
    @PrimaryKey
    @ColumnInfo(name = "uuid") val uuid: String,

    @ColumnInfo(name = "user_id") val userId: Int,

    val date: String,                                       // "YYYY-MM-DD"
    @ColumnInfo(name = "bucket_start") val bucketStart: String,  // "HH:MM" début de tranche
    val steps: Int,

    @ColumnInfo(defaultValue = "0") val synced: Boolean = false,
    @ColumnInfo(defaultValue = "0") val pendingDeletion: Boolean = false,

    @ColumnInfo(name = "updated_at") val updatedAt: String? = getNowISO8601(),
)

File: serveur/app/models/health_step_count.py
# app/models/health_step_count.py
#
# Santé / Health Connect V1 (2026-06-17). Type A user-scoped.
# Compteur de pas en buckets intraday : chaque row = une tranche de la journée
# (ex. tranche horaire). Le total quotidien se dérive par SUM(steps) sur la date.
# Le bucketing permet le near-real-time : le client ré-upsert le bucket courant
# au fil de la journée (uuid stable par user+date+bucket) sans attendre la fin de
# journée. Granularité (tranche) décidée par le client, pas figée en base.

from sqlalchemy import Column, Integer, String, ForeignKey, DateTime
from app.database import Base
import uuid


class HealthStepCount(Base):
    __tablename__ = "health_step_counts"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    uuid = Column(String, nullable=False, unique=True, default=lambda: str(uuid.uuid4()))

    date = Column(String, nullable=False, index=True)   # "YYYY-MM-DD" (convention projet)
    bucket_start = Column(String, nullable=False)        # "HH:MM" début de la tranche intraday
    steps = Column(Integer, nullable=False)

    updated_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return (
            f"HealthStepCount(id={self.id}, user_id={self.user_id}, uuid='{self.uuid}', "
            f"date='{self.date}', bucket_start='{self.bucket_start}', steps={self.steps})"
        )
